File: src/bigclaw/risk.py
from dataclasses import dataclass, field
from typing import Dict, List, Optional

from .models import Priority, RiskLevel, Task


@dataclass
class BudgetDecision:
    status: str
    estimated_cost: float
    remaining_budget: float
    reason: str


class CostController:
    def __init__(self, medium_hourly_costs: Optional[Dict[str, float]] = None):
        self.medium_hourly_costs = medium_hourly_costs or {
            "docker": 2.0,
            "browser": 4.0,
            "vm": 8.0,
            "none": 0.0,
        }

    def estimate_cost(self, medium: str, duration_minutes: int) -> float:
        hourly = self.medium_hourly_costs.get(medium, 0.0)
        return round(hourly * (max(0, duration_minutes) / 60.0), 2)

    def evaluate(self, task: Task, medium: str, duration_minutes: int, spent_so_far: float = 0.0) -> BudgetDecision:
        estimated = self.estimate_cost(medium, duration_minutes)
        effective_budget = float(task.budget + task.budget_override_amount)
        remaining = round(effective_budget - spent_so_far - estimated, 2)

        if effective_budget <= 0:
            return BudgetDecision(
                status="allow",
                estimated_cost=estimated,
                remaining_budget=remaining,
                reason="budget not set",
            )

        if remaining >= 0:
            return BudgetDecision(
                status="allow",
                estimated_cost=estimated,
                remaining_budget=remaining,
                reason="within budget",
            )

        downgraded_medium = "docker" if medium in {"browser", "vm"} else "none"
        if downgraded_medium != medium:
            downgraded_estimated = self.estimate_cost(downgraded_medium, duration_minutes)
            downgraded_remaining = round(effective_budget - spent_so_far - downgraded_estimated, 2)
            if downgraded_remaining >= 0:
                return BudgetDecision(
                    status="degrade",
                    estimated_cost=downgraded_estimated,
                    remaining_budget=downgraded_remaining,
                    reason=f"degrade to {downgraded_medium} to stay within budget",
                )

        return BudgetDecision(
            status="pause",
            estimated_cost=estimated,
            remaining_budget=remaining,
            reason="budget exceeded",
        )


@dataclass
class RiskFactor:
    name: str
    points: int
    reason: str


@dataclass
class RiskScore:
    level: RiskLevel
    total: int
    requires_approval: bool
    factors: List[RiskFactor] = field(default_factory=list)

    @property
    def summary(self) -> str:
        return ", ".join(f"{factor.name}={factor.points}" for factor in self.factors) or "baseline=0"


class RiskScorer:
    TOOL_POINTS = {
        "browser": 10,
        "terminal": 15,
        "github": 10,
        "deploy": 20,
        "sql": 15,
        "warehouse": 15,
        "bi": 10,
    }
    LABEL_POINTS = {
        "security": 20,
        "compliance": 20,
        "prod": 20,
        "release": 15,
        "ops": 10,
    }

    def score_task(self, task: Task) -> RiskScore:
        factors: List[RiskFactor] = []
        total = 0

        risk_points = {
            RiskLevel.LOW: 0,
            RiskLevel.MEDIUM: 30,
            RiskLevel.HIGH: 60,
        }[task.risk_level]
        total += risk_points
        if risk_points:
            factors.append(RiskFactor("risk-level", risk_points, f"declared risk level {task.risk_level.value}"))

        if task.priority == Priority.P0:
            total += 10
            factors.append(RiskFactor("priority", 10, "p0 task needs tighter controls"))

        labels = {label.lower() for label in task.labels}
        for label in sorted(labels):
            points = self.LABEL_POINTS.get(label)
            if points:
                total += points
                factors.append(RiskFactor(f"label:{label}", points, f"label {label} increases operational risk"))

        tools = {tool.lower() for tool in task.required_tools}
        for tool in sorted(tools):
            points = self.TOOL_POINTS.get(tool)
            if points:
                total += points
                factors.append(RiskFactor(f"tool:{tool}", points, f"tool {tool} expands execution surface"))

        if task.budget < 0:
            total += 20
            factors.append(RiskFactor("budget", 20, "invalid budget requires manual review"))

        level = self._level_for_total(total)
        return RiskScore(level=level, total=total, requires_approval=(level == RiskLevel.HIGH), factors=factors)

    def _level_for_total(self, total: int) -> RiskLevel:
        if total >= 60:
            return RiskLevel.HIGH
        if total >= 25:
            return RiskLevel.MEDIUM
        return RiskLevel.LOW

File: bigclaw-go/internal/risk/risk.go
package risk

import (
	"fmt"
	"sort"
	"strconv"
	"strings"

	"bigclaw-go/internal/domain"
)

type Factor struct {
	Name   string `json:"name"`
	Points int    `json:"points"`
	Reason string `json:"reason"`
}

type Score struct {
	Level            domain.RiskLevel `json:"level"`
	Total            int              `json:"total"`
	RequiresApproval bool             `json:"requires_approval"`
	Summary          string           `json:"summary"`
	Factors          []Factor         `json:"factors,omitempty"`
}

var toolPoints = map[string]int{
	"browser":   10,
	"terminal":  15,
	"shell":     15,
	"github":    10,
	"deploy":    20,
	"sql":       15,
	"warehouse": 15,
	"bi":        10,
	"vm":        15,
}

var labelPoints = map[string]int{
	"security":   20,
	"compliance": 20,
	"prod":       20,
	"release":    15,
	"ops":        10,
}

var codeImpactPoints = map[string]int{
	"medium":   10,
	"high":     20,
	"critical": 25,
}

func ScoreTask(task domain.Task, events []domain.Event) Score {
	factors := make([]Factor, 0)
	total := 0
	add := func(name string, points int, reason string) {
		if points <= 0 {
			return
		}
		total += points
		factors = append(factors, Factor{Name: name, Points: points, Reason: reason})
	}

	switch task.RiskLevel {
	case domain.RiskMedium:
		add("risk_level", 30, "declared risk level medium")
	case domain.RiskHigh:
		add("risk_level", 60, "declared risk level high")
	}

	switch {
	case task.Priority > 0 && task.Priority <= 1:
		add("priority", 10, "p0 task needs tighter controls")
	case task.Priority == 2:
		add("priority", 5, "high-priority task increases operational pressure")
	}

	labels := make([]string, 0)
	seenLabels := make(map[string]struct{})
	for _, label := range task.Labels {
		normalized := strings.ToLower(strings.TrimSpace(label))
		if normalized == "" {
			continue
		}
		if _, ok := seenLabels[normalized]; ok {
			continue
		}
		seenLabels[normalized] = struct{}{}
		labels = append(labels, normalized)
	}
	sort.Strings(labels)
	for _, label := range labels {
		if points := labelPoints[label]; points > 0 {
			add("label:"+label, points, fmt.Sprintf("label %s increases operational risk", label))
		}
	}

	tools := make([]string, 0)
	seenTools := make(map[string]struct{})
	for _, tool := range task.RequiredTools {
		normalized := strings.ToLower(strings.TrimSpace(tool))
		if normalized == "" {
			continue
		}
		if _, ok := seenTools[normalized]; ok {
			continue
		}
		seenTools[normalized] = struct{}{}
		tools = append(tools, normalized)
	}
	sort.Strings(tools)
	for _, tool := range tools {
		if points := toolPoints[tool]; points > 0 {
			add("tool:"+tool, points, fmt.Sprintf("tool %s expands execution surface", tool))
		}
	}

	if codeImpact := metadataString(task, "code_impact", "impact_scope", "change_scope"); codeImpact != "" {
		normalized := strings.ToLower(codeImpact)
		if points := codeImpactPoints[normalized]; points > 0 {
			add("code_impact", points, fmt.Sprintf("code impact %s expands blast radius", normalized))
		}
	}

	if task.BudgetCents < 0 {
		add("budget", 20, "invalid budget requires manual review")
	}

	if changedFiles := metadataInt(task, "changed_files_count", "files_changed_count", "code_files_changed"); changedFiles >= 50 {
		add("changed_files", 20, "large code delta expands review scope")
	} else if changedFiles >= 10 {
		add("changed_files", 10, "multi-file change expands review scope")
	}

	failureCount := metadataInt(task, "failure_count", "historical_failures") + countFailureEvents(events)
	if failureCount > 0 {
		points := failureCount * 5
		if points > 15 {
			points = 15
		}
		add("failure_history", points, fmt.Sprintf("%d prior failures on this task/run", failureCount))
	}

	retryCount := metadataInt(task, "retry_count", "historical_retries") + countEventType(events, domain.EventTaskRetried)
	if retryCount > 0 {
		points := retryCount * 5
		if points > 15 {
			points = 15
		}
		add("retry_history", points, fmt.Sprintf("%d retries indicate unstable execution", retryCount))
	}

	regressionCount := metadataInt(task, "regression_count", "historical_regressions")
	if regressionCount > 0 {
		points := regressionCount * 10
		if points > 20 {
			points = 20
		}
		add("regression_history", points, fmt.Sprintf("%d regressions require tighter oversight", regressionCount))
	}

	explicitApproval := metadataBool(task, "approval_required", "requires_approval")
	if explicitApproval {
		add("approval_override", 10, "task metadata requires manual approval")
	}

	level := levelForTotal(total)
	requiresApproval := explicitApproval || level == domain.RiskHigh
	return Score{
		Level:            level,
		Total:            total,
		RequiresApproval: requiresApproval,
		Summary:          summarizeFactors(factors),
		Factors:          factors,
	}
}

func levelForTotal(total int) domain.RiskLevel {
	if total >= 60 {
		return domain.RiskHigh
	}
	if total >= 25 {
		return domain.RiskMedium
	}
	return domain.RiskLow
}

func summarizeFactors(factors []Factor) string {
	if len(factors) == 0 {
		return "baseline=0"
	}
	parts := make([]string, 0, len(factors))
	for _, factor := range factors {
		parts = append(parts, fmt.Sprintf("%s=%d", factor.Name, factor.Points))
	}
	return strings.Join(parts, ", ")
}

func countFailureEvents(events []domain.Event) int {
	count := 0
	for _, event := range events {
		switch event.Type {
		case domain.EventTaskDeadLetter, domain.EventTaskCancelled:
			count++
		}
	}
	return count
}

func countEventType(events []domain.Event, want domain.EventType) int {
	count := 0
	for _, event := range events {
		if event.Type == want {
			count++
		}
	}
	return count
}

func metadataString(task domain.Task, keys ...string) string {
	for _, key := range keys {
		if value := strings.TrimSpace(task.Metadata[key]); value != "" {
			return value
		}
	}
	return ""
}

func metadataInt(task domain.Task, keys ...string) int {
	for _, key := range keys {
		if value := strings.TrimSpace(task.Metadata[key]); value != "" {
			if parsed, err := strconv.Atoi(value); err == nil {
				return parsed
			}
		}
	}
	return 0
}

func metadataBool(task domain.Task, keys ...string) bool {
	for _, key := range keys {
		if value := strings.TrimSpace(task.Metadata[key]); value != "" {
			if parsed, err := strconv.ParseBool(value); err == nil {
				return parsed
			}
		}
	}
	return false
}
